enclave-runtime/src/tls_ra/tls_ra_server.rs

```diff
@@ -0,0 +1,156 @@
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+//! Implementation of the server part of the state provisioning.
+
+use super::{authentication::ClientAuth, Opcode, TcpHeader};
+use crate::{
+	attestation::create_ra_report_and_signature,
+	error::{Error as EnclaveError, Result as EnclaveResult},
+	ocall::OcallApi,
+	tls_ra::seal_handler::{SealHandler, UnsealStateAndKeys},
+};
+use itp_ocall_api::EnclaveAttestationOCallApi;
+use itp_sgx_crypto::{AesSeal, Rsa3072Seal};
+use itp_stf_state_handler::GlobalFileStateHandler;
+use itp_types::ShardIdentifier;
+use rustls::{ServerConfig, ServerSession, Stream};
+use sgx_types::*;
+use std::{
+	backtrace::{self, PrintFormat},
+	io::{Read, Write},
+	net::TcpStream,
+	sync::Arc,
+};
+/// Server part of the TCP-level connection and the underlying TLS-level session.
+///
+/// Includes a seal handler, which handles the reading part of the data to be sent.
+struct TlsServer<'a, StateAndKeyUnsealer> {
+	tls_stream: Stream<'a, ServerSession, TcpStream>,
+	seal_handler: StateAndKeyUnsealer,
+}
+
+impl<'a, StateAndKeyUnsealer> TlsServer<'a, StateAndKeyUnsealer>
+where
+	StateAndKeyUnsealer: UnsealStateAndKeys,
+{
+	fn new(
+		tls_stream: Stream<'a, ServerSession, TcpStream>,
+		seal_handler: StateAndKeyUnsealer,
+	) -> Self {
+		Self { tls_stream, seal_handler }
+	}
+
+	/// Sends all relevant data of the specific shard to the client.
+	fn write_shard(&mut self) -> EnclaveResult<()> {
+		let shard = self.read_shard()?;
+		self.write_all(&shard)
+	}
+
+	/// Read the shard of the state the client wants to receive.
+	fn read_shard(&mut self) -> EnclaveResult<ShardIdentifier> {
+		let mut shard_holder = ShardIdentifier::default();
+		let shard = shard_holder.as_fixed_bytes_mut();
+		self.tls_stream.read(shard)?;
+		Ok(shard.into())
+	}
+
+	/// Sends all relevant data to the client.
+	fn write_all(&mut self, shard: &ShardIdentifier) -> EnclaveResult<()> {
+		let shielding_key = self.seal_handler.unseal_shielding_key()?;
+		let signing_key = self.seal_handler.unseal_signing_key()?;
+		let state = self.seal_handler.unseal_state(shard)?;
+		self.write(Opcode::ShieldingKey, &shielding_key)?;
+		self.write(Opcode::SigningKey, &signing_key)?;
+		self.write(Opcode::State, &state)?;
+		Ok(())
+	}
+
+	/// Sends the header followed by the payload.
+	fn write(&mut self, opcode: Opcode, bytes: &[u8]) -> EnclaveResult<()> {
+		self.write_header(TcpHeader::new(opcode, bytes.len() as u64))?;
+		self.tls_stream.write(bytes)?;
+		Ok(())
+	}
+
+	/// Sends the header which includes the payload length and the Opcode indicating the payload type.
+	fn write_header(&mut self, tcp_header: TcpHeader) -> EnclaveResult<()> {
+		self.tls_stream.write(&tcp_header.opcode.to_bytes())?;
+		self.tls_stream.write(&tcp_header.payload_length.to_be_bytes())?;
+		Ok(())
+	}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn run_state_provisioning_server(
```

Comment:
Code from here on (in this file) left as is (copy pasted from tls_ra.rs), except the inclusion of the `run_state_provisioning_server_internal` for easy testing & usage of `?`